gtsam/hybrid/tests/testGaussianMixtureFactor.cpp

```diff
@@ -191,8 +198,323 @@ TEST(GaussianMixtureFactor, Error) {
+
+  VectorValues given;
+  given.insert(z0, Vector1(0.5));
+  // The motion model measurement says we didn't move
+  given.insert(f01, Vector1(0.0));
+
+  {
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since no measurement on x1, we hedge our bets
+    DiscreteConditional expected(m1, "0.5/0.5");
+    // regression
+    EXPECT(assert_equal(expected, *(bn->at(2)->asDiscrete())));
+  }
+
+  {
+    // Now we add a measurement z1 on x1
+    hbn = CreateBayesNet(mu0, mu1, sigma, sigma, true);
+
+    // If we see z1=2.2, discrete mode should say m1=1
+    given.insert(z1, Vector1(2.2));
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since we have a measurement on z2, we get a definite result
+    DiscreteConditional expected(m1, "0.4923083/0.5076917");
+    // regression
+    EXPECT(assert_equal(expected, *(bn->at(2)->asDiscrete()), 1e-6));
+  }
+}
+
+/* ************************************************************************* */
+/**
+ * Test a model P(x0)P(z0|x0)P(f01|x1,x0,m1)P(z1|x1)P(m1).
+ *
+ * P(f01|x1,x0,m1) has different means and different covariances.
+ *
+ * Converting to a factor graph gives us
+ * P(x0)ϕ(x0)ϕ(x1,x0,m1)ϕ(x1)P(m1)
+ *
+ * If we only have a measurement on z0, then
+ * the P(m1) should be 0.5/0.5.
+ * Getting a measurement on z1 gives use more information.
+ */
+TEST(GaussianMixtureFactor, TwoStateModel2) {
+  using namespace test_two_state_estimation;
+
+  double mu0 = 1.0, mu1 = 3.0;
+  double sigma0 = 6.0, sigma1 = 4.0;
+  auto model0 = noiseModel::Isotropic::Sigma(1, sigma0);
+  auto model1 = noiseModel::Isotropic::Sigma(1, sigma1);
+
+  DiscreteKey m1(M(1), 2);
+  Key z0 = Z(0), z1 = Z(1), f01 = F(0);
+
+  // Start with no measurement on x1, only on x0
+  HybridBayesNet hbn = CreateBayesNet(mu0, mu1, sigma0, sigma1, false);
+
+  VectorValues given;
+  given.insert(z0, Vector1(0.5));
+  // The motion model measurement says we didn't move
+  given.insert(f01, Vector1(0.0));
+
+  {
+    // Start with no measurement on x1, only on x0
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since no measurement on x1, we a 50/50 probability
+    auto p_m = bn->at(2)->asDiscrete();
+    EXPECT_DOUBLES_EQUAL(0.5, p_m->operator()(DiscreteValues{{m1.first, 0}}),
+                         1e-9);
+    EXPECT_DOUBLES_EQUAL(0.5, p_m->operator()(DiscreteValues{{m1.first, 1}}),
+                         1e-9);
+  }
+
+  {
+    // Now we add a measurement z1 on x1
+    hbn = CreateBayesNet(mu0, mu1, sigma0, sigma1, true);
+
+    given.insert(z1, Vector1(2.2));
+    HybridGaussianFactorGraph gfg = hbn.toFactorGraph(given);
+    HybridBayesNet::shared_ptr bn = gfg.eliminateSequential();
+
+    // Since we have a measurement on z2, we get a definite result
```

Comment:
Indicate regression. Would be even better if we could calculate the correct result. Probably you could even use your sigmoid function above, after manually eliminating a bit? Not sure.

Reply:
Yup! I did this for the GaussianMixture2 test. Used `prob_m_z` (the new name for the sigmoid function).